packages/ts/react-crud/src/autoform.tsx

```diff
@@ -17,23 +17,118 @@ type SubmitEvent<TItem> = {
   disabled?: boolean;
+  customLayoutRenderer?: AutoFormLayoutProps | ComponentType<AutoFormLayoutRendererProps<M>>;
   onSubmitError?({ error }: SubmitErrorEvent): void;
-  afterSubmit?({ item }: SubmitEvent<TItem>): void;
+  afterSubmit?({ item }: SubmitEvent<Value<M>>): void;
+}>;
+
+type CustomFormLayoutProps = Readonly<{
+  customFormLayout: AutoFormLayoutProps;
+  children: ReadonlyArray<ReactElement<AutoFormFieldProps>>;
 }>;
 
-export function ExperimentalAutoForm<TItem>({
+function findColumnCount(responsiveSteps: Array<{ minWidth: string; columns: number }>): number {
+  return responsiveSteps
+    .map((step: { minWidth: string; columns: number }) => ({
+      minWidth: parseInt(step.minWidth, 10),
+      columns: step.columns,
+    }))
+    .filter(({ minWidth }) => minWidth <= window.innerWidth)
+    .reduce(
+      (maxStep, step) => {
+        if (step.minWidth > maxStep.minWidth) {
+          return step;
+        }
+        return maxStep;
+      },
+      { minWidth: 0, columns: 1 },
+    ).columns;
+}
+
+function createGenericResponsiveSteps(
+  customFormLayout: AutoFormLayoutProps,
+): Array<{ minWidth: string; columns: number }> {
+  function gcd(a: number, b: number): number {
+    return a > 0 ? gcd(b % a, a) : b;
+  }
+  const lcm = (a: number, b: number) => (a * b) / gcd(a, b);
+
+  const minNeededColumns = customFormLayout.template
+    .map((row) => row.length)
+    .filter((value, index, array) => array.indexOf(value) === index)
+    .reduce(lcm);
+
+  return [
+    { minWidth: '0', columns: 1 },
+    { minWidth: '800px', columns: minNeededColumns },
+  ];
+}
+
+function CustomFormLayout(customFormLayoutProps: CustomFormLayoutProps): JSX.Element {
+  const { customFormLayout, children } = customFormLayoutProps;
+  const fieldsByPropertyName = new Map<string, JSX.Element>();
+  children.forEach((field) => fieldsByPropertyName.set(field.props.propertyInfo.name, field));
+
+  let responsiveSteps: Array<{ minWidth: string; columns: number }>;
+  if (customFormLayout.responsiveSteps == null) {
+    responsiveSteps = createGenericResponsiveSteps(customFormLayout);
+  } else {
+    ({ responsiveSteps } = customFormLayout);
+  }
+
+  let weightedTemplate: FieldColSpan[][];
+  if (typeof customFormLayout.template[0][0] === 'string') {
+    const columnCount = findColumnCount(responsiveSteps);
+    weightedTemplate = (customFormLayout.template as string[][]).map((row) => {
+      const rowSize = row.length;
+      const colSpan = Math.ceil(columnCount / rowSize);
+      return row.map((property) => ({ property, colSpan }));
+    });
+  } else {
+    weightedTemplate = customFormLayout.template as FieldColSpan[][];
+  }
+
+  const spannedFields: JSX.Element[] = [];
+  weightedTemplate.forEach((row: FieldColSpan[]) => {
+    row.forEach((fieldColSpan: FieldColSpan) => {
+      const field = fieldsByPropertyName.get(fieldColSpan.property)!;
+      const spannedField = React.cloneElement(field, { colspan: fieldColSpan.colSpan });
```

Comment:
This seems to be discouraged. Did you try one of the [alternatives](https://react.dev/reference/react/cloneElement#alternatives)?

Reply:
Done.